rasa/core/processor.py

```diff
@@ -829,15 +829,26 @@ def _log_action_on_tracker(
             tracker.update_with_events(prediction.events, self.domain)
 
             # log the action and its produced events
-            tracker.update(
-                ActionExecuted(
-                    action_name, prediction.policy_name, prediction.max_confidence
-                )
-            )
+            tracker.update(self._action_event_for_prediction(action_name, prediction))
 
         logger.debug(f"Action '{action_name}' ended with events '{events}'.")
         tracker.update_with_events(events, self.domain)
 
+    def _action_event_for_prediction(
+        self, action_name: Text, prediction: PolicyPrediction
+    ) -> ActionExecuted:
+        action_text = None
+        if action_name in self.domain.action_texts:
```

Comment:
I would prefer modifying the part of the code (maybe inside the `domain.py`) that transforms predicted index into an action name, to return a tuple `(action_name, action_text)` one of which would be `None`. Then all of these method should take both name and text and pass it along. this way we decouple the logic of text vs name and contain it only within domain file

Reply:
We only pass around the `Action` object which we retrieve [here](https://github.com/RasaHQ/rasa/blob/d43364627ce27b6c7048a7e460a0087a84447f3e/rasa/core/processor.py#L357). I think passing around additional information via tuples rather increases complexity than to help with it. What I could do is to set a flag `is_end_to_end_action` on the `Action` object [here](https://github.com/RasaHQ/rasa/blob/8f60f83d089e0e4dcbd79900c7598935f50119e6/rasa/core/actions/action.py#L173) or even add method `action_executed_for_event` to `Action` which returns the correct `ActionExecuted` event.

Reply:
I think `action_for_index` is a perfect place for that. Instead of doing it inside `action_for_name`, create method `action_for_text` and differentiate between them inside `action_for_index`. What do you think?

Reply:
I don't see any advantage of doing this. What are you to accomplish? If you want to isolate determining whether it's an end-to-end action, then I'd go for the described approach above.

Reply:
because the name `action_for_name` is confusing. Name doesn't decide the action anymore, it is the index that decides that. Why do we have 2 identical methods on the first place?

Reply:
I want to avoid putting `action_text` into variable called `action_name` in a couple of month it'll create terrible confusion

Reply:
> Why do we have 2 identical methods on the first place?

Which two methods are you referring to? Index and name? There are some cases where have a name and want to get the action based on that (e.g. for the `action_session_started)

> I want to avoid putting action_text into variable called action_name in a couple of month it'll create terrible confusion

I see that. As only `_log_action_on_tracker` is having this variable I still suggest doing what I suggested [in my earlier. comment.](https://github.com/RasaHQ/rasa/pull/7325#discussion_r530490382).

Reply:
I'm not sure I fully follow your proposal, but if you think it is better let's go for it. I'd prefer the implementation where there is no need for additional flags

Reply:
I agree on the flag thing and I've an idea about that. Can hopefully do a quick draft later which we can then discuss.

Reply:
@Ghostvv Updated it. Way cleaner solution now. I was able to get rid of all domain changes. Haven't polished the PR yet because I wanted to share it with you first, so please comment on the higher level things